### ISSUE TITLE
Default content-type should not override set headers

### ISSUE THREAD

**Issue by orevial** [flutterchina/dio#1076]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

Might be related to #1045 and #1059 though.

### Issue Info

| Info                           | Value                                                 |    
| ------------------------------ | ----------------------------------------------------- | 
| Platform                  | Flutter 2.0.1  /  Dart 2.12.0                        |      
| Dio Version                    | ^4.0.0-beta6                                  |     

### Issue Description and Steps

`content-type` header (set through `options.headers`) gets overriden by default `application/json` content-type when doing (at least) post requests.

Example:

```dart
client.post<String>(
        'someUrl',
        data: '<xml>data</xml>',
        options: Options(
          headers: <String, String>{
            'content-type': 'application/xml',
            'authorization': 'bearer SOMETOKEN',
          },
        ),
      )
```

Although I would expect my `application/xml` content-type header to be passed in this case, the actual content-type header getting through to the server is **always** `application/json` (no matter if the header name is lower or uppercase by the way).

To make it work we have to set the content-type through the options field:

```dart
client.post<String>(
        'someUrl',
        data: '<xml>data</xml>',
        options: Options(
          contentType: 'application/xml',
          headers: <String, String>{
            'authorization': 'bearer SOMETOKEN',
          },
        ),
      )
```

Although it works fine as a workaround, it does not feel very natural that one would override the other, or it should at least be documented, because it worked OK in previous version of dio (3.0.10).

I would suggest multiple things:
1. Fixing the headers precendence. Maybe take the `contentType` parameter first if it is set, then take a user-set `content-type` header, then - and only then - maybe default to the default `application/json` header
2. Detecting and handling the merge of the `content-type` header no matter the variable case (even if all get later converted to lowercase)
3. Adding an assertion in Options constructor so that it is impossible for the user to set both the `contentType` parameter **and** a `content-type` header, as it is does not make sense (the user can't know for sure which one is to be used by dio). Something like this: 
  ```dart
    Options({ // params here... }) :
         assert(contentType == null || 
          !headers?.keys.map((e) => e.toLowerCase()).contains('content-type') ?? false, 
          'You cannot set both contentType param and a content-type header',);
  ```

**Comment by wendux** [comment]
You are right, I will fix it later.

**Comment by wendux** [comment]
fixed in 4.0.0-prev2
